registry_provider.go

```diff
@@ -0,0 +1,259 @@
+package tfe
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+)
+
+// Compile-time proof of interface implementation.
+var _ RegistryProviders = (*registryProviders)(nil)
+
+// RegistryProviders describes all the registry provider-related methods that the Terraform
+// Enterprise API supports.
+//
+// TFE API docs: https://www.terraform.io/docs/cloud/api/providers.html
+type RegistryProviders interface {
+	// List all the providers within an organization.
+	List(ctx context.Context, organization string, options *RegistryProviderListOptions) (*RegistryProviderList, error)
+
+	// Create a registry provider.
+	Create(ctx context.Context, organization string, options RegistryProviderCreateOptions) (*RegistryProvider, error)
+
+	// Read a registry provider.
+	Read(ctx context.Context, providerID RegistryProviderID, options *RegistryProviderReadOptions) (*RegistryProvider, error)
+
+	// Delete a registry provider.
+	Delete(ctx context.Context, providerID RegistryProviderID) error
+}
+
+// registryProviders implements RegistryProviders.
+type registryProviders struct {
+	client *Client
+}
+
+// RegistryName represents which registry is being targeted
+type RegistryName string
+
+// List of available registry names
+const (
+	PrivateRegistry RegistryName = "private"
+	PublicRegistry  RegistryName = "public"
+)
+
+// RegistryProviderIncludeOps represents which jsonapi include can be used with registry providers
+type RegistryProviderIncludeOps string
+
+// List of available includes
+const (
+	RegistryProviderVersionsInclude RegistryProviderIncludeOps = "registry-provider-versions"
+)
+
+// RegistryProvider represents a registry provider
+type RegistryProvider struct {
+	ID           string                       `jsonapi:"primary,registry-providers"`
+	Name         string                       `jsonapi:"attr,name"`
+	Namespace    string                       `jsonapi:"attr,namespace"`
+	CreatedAt    string                       `jsonapi:"attr,created-at,iso8601"`
+	UpdatedAt    string                       `jsonapi:"attr,updated-at,iso8601"`
+	RegistryName RegistryName                 `jsonapi:"attr,registry-name"`
+	Permissions  RegistryProviderPermissions `jsonapi:"attr,permissions"`
+
+	// Relations
+	Organization             *Organization              `jsonapi:"relation,organization"`
+	RegistryProviderVersions []*RegistryProviderVersion `jsonapi:"relation,registry-provider-versions"`
+
+	// Links
+	Links map[string]interface{} `jsonapi:"links,omitempty"`
+}
+
+type RegistryProviderPermissions struct {
+	CanDelete bool `jsonapi:"attr,can-delete,omitempty"`
+	CanUploadAsset bool `jsonapi:"attr,can-upload-asset,omitempty"`
```

Comment:
@brandonc `CanDelete` is a permission only for the registry provider, and `CanUploadAsset` is only for a registry provider version. Initially I liked having both permissions in one structure, but I could see splitting this up into 2 separate structs, one for RegistryProvider and one for RegistryProviderVersion. What are your thoughts?

Reply:
Also see this [comment](https://github.com/hashicorp/go-tfe/pull/405/files#r876424213) & this [comment](https://github.com/hashicorp/go-tfe/pull/405/files#r876425389)

Reply:
Yeah, that is my preference